🎉 New Source: Tyntec SMS [low-code cdk]

michizhou opened the pull request:
## What
*Developing new connector for the source: Tyntec SMS*
*Documentation: https://api.tyntec.com/reference/sms/current.html#sms-api*

Resolves:
🎉 [New Source: Tyntec SMS](https://github.com/airbytehq/connector-contest/issues/184)

## How
*Developed using Low-Code CDK*

## Recommended reading order
1. `spec.yaml`
2. `tyntec_sms.yaml`
3. `schemas/messages.json`

## 🚨 User Impact 🚨
No breaking changes. Just the addition of a new source.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

marcosmarxm commented:
Hello @michizhou, Marcos from Airbyte here 👋 . We received more than 25 new contributions along the weekend. One is yours 🎉  thank so much for! Our team is limited and maybe the review process can take longer than expected. As described in the Airbyte's Hacktoberfest your contribution was submitted before November 2nd and it is eligible to win the prize. The review process will validate other requirements. I ask to you patience until someone from the team review it. 

Because I reviewed some contributions for Hacktoberfest so far I saw some common patterns you can check in advance:
* [ ] Make sure you have added connector documentation to `/docs/integrations/`
* [ ] Remove the file `catalog` from `/integration_tests`
* [ ] Edit the `sample_config.json` inside `/integration_tests`
* [ ] For the `configured_catalog` you can use only `json_schema: {}`
* [ ] Add title to all properties in the `spec.yaml`
* [ ] Make sure the `documentationUrl` in the `spec.yaml` redirect to Airbyte's future connector page, eg: connector Airtable the `documentationUrl: https://docs.airbyte.com/integrations/sources/airtable`
* [ ] Review now new line at EOF (end-of-file) for all files.

If possible send to me a DM in Slack with the tests credentials, this process will make easier to us run integration tests and publish your connector. If you only has production keys, make sure to create a bootstrap.md explaining how to get the keys.

marcosmarxm commented:
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

michizhou commented:
@marcosmarxm Thanks for taking your time to review this. I have DM'ed you on slack with test credentials and to the best of my knowledge I believe I have fulfilled all the checklist items:

* [x]  Make sure you have added connector documentation to `/docs/integrations/`
* [x]  Remove the file `catalog` from `/integration_tests`
* [x]  Edit the `sample_config.json` inside `/integration_tests`
* [x]  For the `configured_catalog` you can use only `json_schema: {}`
* [x]  Add title to all properties in the `spec.yaml`
* [x]  Make sure the `documentationUrl` in the `spec.yaml` redirect to Airbyte's future connector page, eg: connector Airtable the `documentationUrl: https://docs.airbyte.com/integrations/sources/airtable`
* [x]  Review now new line at EOF (end-of-file) for all files.

marcosmarxm commented on airbyte-integrations/connectors/source-tyntec-sms/source_tyntec_sms/tyntec_sms.yaml:
```diff
@@ -0,0 +1,59 @@
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+  # API Docs: https://api.tyntec.com/reference/sms/current.html#sms-api-Read%20SMS%20status
+  messages_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "messages"
+      primary_key: "doneDate"
+      path: "/messaging/v1/messages/{{ config['request_id'] }}"
```
Why you don't use the other endpoint to collect all `request_id` messgages?
https://api.tyntec.com/reference/sms/current.html#sms-api-Send%20SMS%20(GET)

Are you sure the primary key is `DoneDate`?

michizhou replied:
@marcosmarxm Are you suggesting the linked endpoint should be used to create a new stream to get `request_id`, which would then be passed onto the `messages` stream as a parameter? If so, how could I configure such a stream and the `messages` stream so they link up properly?

As for the primary key being `doneDate`, the response doesn't have a designated key so I picked that as a reasonable choice. Let me know if you think a different field (or even no field) is more suitable from the response fields: https://api.tyntec.com/reference/sms/current.html#sms-api-restmessageresponse

michizhou replied:
@marcosmarxm Are you suggesting the linked endpoint is to be used to create a new stream to get `request_id`, which would then be passed onto the `messages` stream as a parameter? If so, how could I configure such a stream and the `messages` stream so they link up properly? The linked endpoint also seems to be designed to send single SMS messages and return the `request_id` of the sent message, so I don't know if it can retrieve all `request_id` values of previously sent messages. I suppose if we configure it so that every SMS message is sent through this new stream and its `request_id` is stored or immediately used to trigger the `messages` stream, that is a viable alternative.

As for the primary key being `doneDate`, the response doesn't have a designated key so I picked it as a reasonable choice. Let me know if you think a different field (or even no field) is more suitable from the response fields: https://api.tyntec.com/reference/sms/current.html#sms-api-restmessageresponse

michizhou replied:
@marcosmarxm Are you suggesting the linked endpoint is to be used to create a new stream to get `request_id`, which would then be passed onto the `messages` stream as a parameter? If so, how could I set up such a stream and the `messages` stream so they link up properly? The linked endpoint also seems to be designed to send a single SMS message and return the `request_id` of the sent message, so I don't know if it can fetch `request_id` values of previously sent messages. I suppose we can make it work if we configure it so every SMS message is sent via this new stream and the `request_id` is stored or immediately used to trigger the `messages` stream.

As for the primary key being `doneDate`, the response doesn't have a designated key so I picked it as a reasonable choice. Let me know if you think a different field (or even no field) is more suitable from the response fields: https://api.tyntec.com/reference/sms/current.html#sms-api-restmessageresponse

marcosmarxm replied:
Exactly. Please read the docs about StreamSlicer: https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/stream-slicers check other connectors implementation using ListStreamSlicer.

marcosmarxm replied:
Check Survey Sparrow connector, it uses a List StreamSlicer can help you.

marcosmarxm commented on airbyte-integrations/connectors/source-tyntec-sms/source_tyntec_sms/spec.yaml:
```diff
@@ -0,0 +1,21 @@
+      description: Your Tyntec API Key. See <a href="https://www.tyntec.com/docs/docs-center-sms-api-quick-start">here</a>
+      order: 0
+      airbyte_secret: true
+    request_id:
+      type: string
+      title: SMS Message Request ID
+      description: The unique identifier provided for each SMS messaging request.
+      order: 1
```
I think this should be removed and create a new stream to retrieve all requests id/messages and use a stream slices with the message stream.

michizhou replied:
@marcosmarxm The new stream you suggested to retrieve all request IDs/messages will likely require the addition of new fields `to`, `from`, and `message`, which are required for the [linked endpoint](https://api.tyntec.com/reference/sms/current.html#sms-api-Send%20SMS%20(GET)). Is this acceptable?

marcosmarxm replied:
You can implement `to` `from` but check my previous comment.

marcosmarxm reviewed (changes requested):
Some comments.

michizhou commented:
@marcosmarxm I have addressed your comments and sent you a DM on Slack with updated test credentials. Please let me know if any additional changes are needed.

marcosmarxm reviewed (approved):
Thanks @michizhou I tested locally
